File: Frontend/src/lib/validation.ts
export const USERNAME_MIN_LENGTH = 4;
export const USERNAME_MAX_LENGTH = 20;

// Same rules as validate_username in Backend/routes/websocket.py - keep the two
// in sync. The backend check is the real one; this one gives instant feedback.
export function validateUsername(raw: string): string | null {
  const name = raw.trim();
  // Array.from counts characters the way Python's len() does (an emoji is 1, not 2)
  const length = Array.from(name).length;
  // asked for first: "alice smith" should hear about the space, not the length
  if (/\s/.test(name)) {
    return "Username can't contain spaces";
  }
  if (length < USERNAME_MIN_LENGTH) {
    return `Username must be at least ${USERNAME_MIN_LENGTH} characters`;
  }
  if (length > USERNAME_MAX_LENGTH) {
    return `Username must be at most ${USERNAME_MAX_LENGTH} characters`;
  }
  if (!/[0-9]/.test(name)) {
    return 'Username must contain at least one digit';
  }
  return null;
}

export const GROUP_NAME_MIN_LENGTH = 1;
export const GROUP_NAME_MAX_LENGTH = 30;

// Same rules as validate_group_name in Backend/realtime/validation.py - keep the
// two in sync. Looser than a username: no digit is needed, and two groups may
// share a name, because a group name labels a chat rather than a person.
export function validateGroupName(raw: string): string | null {
  const name = raw.trim();
  const length = Array.from(name).length;
  if (/\s/.test(name)) {
    return "Group name can't contain spaces";
  }
  if (length < GROUP_NAME_MIN_LENGTH) {
    return "Group name can't be empty";
  }
  if (length > GROUP_NAME_MAX_LENGTH) {
    return `Group name must be at most ${GROUP_NAME_MAX_LENGTH} characters`;
  }
  return null;
}

File: Backend/realtime/validation.py
import re
import uuid
from realtime.types import AppState, Validate


# Same rules as Frontend/src/lib/validation.ts - keep the two in sync.
def validate_username(username: str) -> str | None:
    """Return why the username breaks the rules, or None if it's valid."""
    username = username.strip()
    # asked for first: "alice smith" should hear about the space, not the length
    if re.search(r"\s", username):
        return "Username can't contain spaces"
    if len(username) < Validate.USERNAME_MIN_LENGTH:
        return f"Username must be at least {Validate.USERNAME_MIN_LENGTH} characters"
    if len(username) > Validate.USERNAME_MAX_LENGTH:
        return f"Username must be at most {Validate.USERNAME_MAX_LENGTH} characters"
    if not re.search(r"[0-9]", username):
        return "Username must contain at least one digit"
    return None


def is_name_taken(state: AppState, username: str, ignore_user_id: uuid.UUID | None = None) -> bool:
    """Is someone already using this name? Pure - it only reads the state it's given.

    ignore_user_id skips one person, so renaming yourself from riya1 to Riya1
    isn't refused as taken by you.
    """
    # casefold() ignores case, so Riya1 and riya1 count as the same name
    return any(
        user.user_name.casefold() == username.casefold()
        for user_id, user in state.users.items()
        if user_id != ignore_user_id
    )


# Same rules as validateGroupName in Frontend/src/lib/validation.ts - keep the two in sync.
def validate_group_name(name: str) -> str | None:
    """Return why the group name breaks the rules, or None if it's valid.

    Looser than a username: no digit is needed, and two groups may share a name.
    A group name labels a conversation; it isn't an identity anyone is addressed by.
    """
    trimmed = name.strip()
    if re.search(r"\s", trimmed):
        return "Group name can't contain spaces"
    if len(trimmed) < Validate.GROUP_NAME_MIN_LENGTH:
        return "Group name can't be empty"
    if len(trimmed) > Validate.GROUP_NAME_MAX_LENGTH:
        return f"Group name must be at most {Validate.GROUP_NAME_MAX_LENGTH} characters"
    return None
